File: go/problems/pg42579/solution.go
// Package pg42579 - [PG] 42579 - 베스트앨범
// https://programmers.co.kr/learn/courses/30/lessons/42579
// 난이도: lv2
// 태그: hash
//
// 시간복잡도: O(NLogN)
// 공간복잡도: O(N)
package pg42579

import "sort"

type song struct {
	play  int
	index int
}
type genre struct {
	name string
	play int
}

func Solution(genres []string, plays []int) []int {
	genresToSum := make(map[string]int)
	genresToSongs := make(map[string][]song)
	for i, play := range plays {
		genresToSum[genres[i]] += play
		genresToSongs[genres[i]] = append(genresToSongs[genres[i]], song{play: play, index: i})
	}

	// 정렬
	sortedGenres := make([]genre, 0, len(genresToSum))
	for key, value := range genresToSum {
		sortedGenres = append(sortedGenres, genre{name: key, play: value})
	}
	sort.Slice(sortedGenres, func(i, j int) bool {
		return sortedGenres[j].play < sortedGenres[i].play
	})

	var answers []int
	// 조회
	for _, element := range sortedGenres {
		songs := genresToSongs[element.name]
		sort.Slice(songs, func(i, j int) bool {
			if songs[i].play != songs[j].play {
				return songs[j].play < songs[i].play
			}
			if songs[i].index != songs[j].index {
				return songs[i].index < songs[j].index
			}
			return true
		})
		limit := 2
		if len(songs) < 2 {
			limit = len(songs)
		}
		for _, s := range songs[:limit] {
			answers = append(answers, s.index)
		}
	}

	return answers
}

File: java/src/main/java/problems/pg42579/Solution.java
package problems.pg42579;

import java.util.*;

/**
 * [PG] 42579 - 베스트앨범 https://programmers.co.kr/learn/courses/30/lessons/42579 난이도: lv2 태그: hash
 *
 * <p>시간복잡도: O(NLogN) 공간복잡도: O(N)
 */
public class Solution {

  public int[] solution(String[] genres, int[] plays) {
    // 정의하기
    Map<String, Integer> genreToTotalPlays = new HashMap<>();
    Map<String, List<Song>> genreToSongs = new HashMap<>();
    for (int i = 0; i < genres.length; i++) {
      genreToTotalPlays.put(genres[i], genreToTotalPlays.getOrDefault(genres[i], 0) + plays[i]);
      genreToSongs.computeIfAbsent(genres[i], k -> new ArrayList<>()).add(new Song(plays[i], i));
    }

    // 장르 정렬: 오름차순으로
    List<String> sortedGenres = new ArrayList<>(genreToTotalPlays.keySet());
    sortedGenres.sort((g1, g2) -> genreToTotalPlays.get(g2) - genreToTotalPlays.get(g1));
    List<Integer> answer = new ArrayList<>();
    for (String genre : sortedGenres) {
      final List<Song> songs = genreToSongs.get(genre);
      songs.sort((s1, s2) -> s1.play == s2.play ? s1.index - s2.index : s2.play - s1.play);
      for (int i = 0; i < Math.min(2, songs.size()); i++) {
        answer.add(songs.get(i).index);
      }
    }
    return answer.stream().mapToInt(Integer::intValue).toArray();
  }

  private record Song(int play, int index) {}
}
